src/libsyntax/print/pp.rs

```diff
@@ -580,10 +583,25 @@ impl<'a> Printer<'a> {
         debug!("print String({})", s);
         // assert!(len <= space);
         self.space -= len;
-        while self.pending_indentation > 0 {
-            write!(self.out, " ")?;
-            self.pending_indentation -= 1;
+
+        // Write the pending indent. A more concise way of doing this would be:
+        //
+        //   write!(self.out, "{: >n$}", "", n = self.pending_indentation as usize)?;
+        //
+        // But that is significantly slower than using `SPACES`. This code is
+        // sufficiently hot, and indents can get sufficiently large, that the
+        // difference is significant on some workloads.
+        let spaces_len = SPACES.len() as isize;
+        while self.pending_indentation >= spaces_len {
+            debug_assert_eq!(SPACES.len() as isize, spaces_len);
+            write!(self.out, "{}", SPACES)?;
+            self.pending_indentation -= spaces_len;
         }
+        if self.pending_indentation > 0 {
+            write!(self.out, "{}", &SPACES[0..self.pending_indentation as usize])?;
```

Comment:
If you want to speed up this code, you could use `self.out.write_str` instead, which bypasses the formatting infrastructure.

Reply:
It's `self.out.write`, but yes, I will do that.